external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -1645,6 +1645,75 @@ fn binop(
             (Type_::bool(loc), operand_ty)
         }
 
+        Eq | Neq if context.env.supports_feature(context.current_package(), FeatureGate::AutoborrowEq) => {
+            let lhs_type = core::ready_tvars(&context.subst, el.ty.clone());
+            let rhs_type = core::ready_tvars(&context.subst, er.ty.clone());
+            let (lhs_ref, lhs_inner, rhs_ref, rhs_inner) = match (lhs_type, rhs_type) {
+                (sp!(_, Type_::Ref(lhs_mut, lhs)), sp!(_, Type_::Ref(rhs_mut, rhs))) => {
+                    (Some(lhs_mut), *lhs, Some(rhs_mut), *rhs)
+                }
+                (sp!(_, Type_::Ref(lhs_mut, lhs)), rhs) => (Some(lhs_mut), *lhs, None, rhs),
+                (lhs, sp!(_, Type_::Ref(rhs_mut, rhs))) => (None, lhs, Some(rhs_mut), *rhs),
+                (lhs, rhs) => (None, lhs, None, rhs),
+            };
+            let ty = join(context, bop.loc, msg, lhs_inner.clone(), rhs_inner.clone());
+            context.add_single_type_constraint(loc, msg(), ty.clone());
+            let eq_ty = match (lhs_ref, rhs_ref) {
+                (None, None) => {
+                    let ability_msg = Some(format!(
+                        "'{}' requires the '{}' ability as the value is consumed. Try \
+                                 borrowing the values with '&' first.'",
+                        &bop,
+                        Ability_::Drop,
+                    ));
+                    context.add_ability_constraint(
+                        el.exp.loc,
+                        ability_msg.clone(),
+                        lhs_inner,
+                        Ability_::Drop,
+                    );
+                    context.add_ability_constraint(
+                        er.exp.loc,
+                        ability_msg,
+                        rhs_inner,
+                        Ability_::Drop,
+                    );
+                    ty
+                }
+                (None, Some(rhs_mut)) => {
+                    // If the RHS is mutable, we reborrow it as immutable so avoid requiring the
+                    // LHS to be mutable.
+                    if rhs_mut {
+                        er = exp_to_ref(context, er.exp.loc, false, er);
+                    }
```

Comment:
Why would we require the LHS to be mutable here? Wouldn't `join` allow that?

Reply:
Whoops, I didn't push the last changes that make this all much nicer / clearer. Hopefully the revision makes everything clear.